### PR TITLE
Prevent duplicated IDs of Module Types and Templates provided by diff…

### PR DESCRIPTION
**Pull request by adimova**
…erent providers

Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

solves #3195

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 1]
```diff
@@ -256,9 +256,19 @@ protected void importFile(String parserType, URL url) {
...
         if (providedObjects != null && !providedObjects.isEmpty()) {
+            List<String> previousPortfolio = providerPortfolio.get(url);
             List<String> uids = new ArrayList<String>();
             for (E providedObject : providedObjects) {
                 String uid = getUID(providedObject);
+                if (providedObjectsHolder.get(uid) == null) {
+                    if (checkExistence(uid)) {
+                        continue;
+                    }
+                } else if (previousPortfolio == null || !previousPortfolio.contains(uid)) {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: Either this is a bug or the code is hard to read - wouldn't you have a problem IF the previousPortfolio already contains this uid?

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 2]
```diff
@@ -256,9 +256,19 @@ protected void importFile(String parserType, URL url) {
...
+                    logger.error("{} with UID \"{}\" already exists! Failed to create a second with the same UID!",
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: In what kind of situation can we get here? If we have multiple files that contain entities with the same uid?
"error" logging should only be used in irrecoverable situations - I would think that "warning" should be enough here to tell the user that he has some problematic files and thus the second entity is simply ignored.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java [hunk 3]
```diff
@@ -49,8 +60,18 @@ protected String getUID(ModuleType providedObject) {
...
-            return Collections.<T>emptyList();
+            return Collections.<T> emptyList();
         }
         return (Collection<T>) new LinkedList<ModuleType>(values);
     }
+
+    @Override
+    protected boolean checkExistence(String uid) {
+        if (moduleTypeRegistry.get(uid) != null) {
+            logger.error("Module Type with UID \"{}\" already exists! Failed to create a second with the same UID!",
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java]: Having an error logging here is very unexpected. A method `checkExistence` should simply return true or false and not have any side effects.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java [hunk 4]
```diff
@@ -53,4 +58,30 @@ public RuleTemplate getTemplate(String UID, Locale locale) {
...
+     * UIDs before these objects to be added in the system.
+     *
+     * @param uid UID of the newly created {@link Template}, which to be checked.
+     * @return {@code true} if {@link Template} with the same UID exists or {@code false} in the opposite
+     *         case.
+     */
+    @Override
+    protected boolean checkExistence(String uid) {
+        if (templateRegistry != null && templateRegistry.get(uid) != null) {
+            logger.error("Rule Template with UID \"{}\" already exists! Failed to create a second with the same UID!",
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java]: Having an error logging here is very unexpected. A method `checkExistence` should simply return true or false and not have any side effects.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 1 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: This verification tells us that an object with such an identifier already exists, but has been added by another file, which could be a problem. IF the previousPortfolio already contains this uid, the object will simply be updated

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java (hunk 2 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: ok

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java (hunk 3 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java]: Will be moved here:
if (checkExistence(uid)) {
      logger.error("Module Type with UID \"{}\" already exists! Failed to create a second with the same UID!"...
      continue;
}

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java (hunk 4 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java]: Will be moved here:
if (checkExistence(uid)) {
logger.error("Module Type with UID "{}" already exists! Failed to create a second with the same UID!"...
continue;
}

**Comment by kaikreuzer** [comment]
Sorry, I just took a step back here and think the overall approach of this PR is wrong.
If I am not mistaken, this would add a cyclic dependency between ModuleTypeFileProvider and ModuleTypeRegistry as well as the TemplateFileProvider and the TemplateRegistry. Or am I missing something?

**Comment by adimova** [comment]
#this would add a cyclic dependency between ModuleTypeFileProvider and ModuleTypeRegistry as well as the TemplateFileProvider and the TemplateRegistry

yes, this is correct, see @danchom comment on  #3195

**Comment by kaikreuzer** [comment]
Thanks for the pointer, @adimova - so I assume we should close this PR and maybe just do some hardening/safeguards in the registries?

**Comment by adimova** [comment]
> just do some hardening/safeguards in the registries?
I agree with that

**Comment by adimova** [comment]
This PR is replaced by #3670

**Pull request closed without merging**